1-Draft/RFCNNNN-Notification-On-Version-Update.md

```diff
@@ -0,0 +1,244 @@
+
+5. After a new update is detected during a successful check,
+all subsequent interactive sessions of that version of `pwsh` should show the notification at startup time.
+And subsequent checks can be avoided for a reasonable period of time, such as a week.
+
+6. `pwsh` of preview versions should check for the new preview version as well as the new GA version.
+`pwsh` of GA versions should check for the new GA version only.
+
+7. The notification and update check are not needed in some scenarios,
+such as when `pwsh` is in a container image.
+Hence, you should be able to suppress them altogether by setting an environment variable.
+
+### No Goals
+
+1. Notification shows up right after a new version of `pwsh` is released.
+
+   _This is not a goal._
+   Assuming you use `pwsh` interactively every day,
+   then a notification about the new release may show up on the same day,
+   but is guaranteed no later than the next day.
+
+2. If an update check detects a new release, the notification should show up in the same session.
+
+   _This is not a goal._
+   An update check should happen way after the startup of an interactive session,
+   and thus it has no impact on whether or not a notification will be shown at the startup of that session.
+   If new release is detected,
+   the subsequent interactive sessions will show a notification about that new release.
+
+### Implementation
+
+This section talks about
+
+- when to do the update check
+- how to persist the detected new release for subsequent `pwsh` sessions to use
+- how to synchronize update checks from different processes of the same version `pwsh` so that at most only one can run to complete during a day
+- how to do the update check
+- how to display the notification
+
+#### When to do the update check
+
+During the startup, `pwsh` creates a `Task` of the update check work,
+but delays the task run for 3 seconds by using `Task.Delay(3000)`.
+The typical startup time for `pwsh` with a moderate size profile should be less than 1 second.
+Given that, I guess it's reasonable to delay the update check work for 3 seconds,
+so that it has close-to-zero impact on the startup performance.
+
+#### How to persist information about a new version
+
+The version of new release is persisted using a file,
+not as the file content, but instead baked in the file name in the following template,
+so that we can avoid extra file loading at the startup.
+
+```none
+_update_<new-version>
+```
+
+The file should be in a folder that is unique to the specific version of `pwsh`.
+For example, for the `v6.2.0 pwsh`, the folder `6.2.0` will be created in the `pwsh` cache folder (shown below),
+and the update check related files for that version of `pwsh` are put there exclusively.
+In this way, the update information for different versions of `pwsh` doesn't interfere with each other.
+
+- Windows: `$env:LOCALAPPDATA\Microsoft\PowerShell\6.2.0`
```

Comment:
Isn't Major.Minor sufficient?

Reply:
We have service releases, like 6.1.3.

Reply:
But we don't have a 6.1.3.x release, so keeping the data in `6.1` would be sufficient.  Even if you have 6.1.1 and 6.1.3 side-by-side, they should share the update information and inform the user to update to 6.1.4, for example.

Reply:
The folder name comes from `$PSVersionTable.GitCommitId`.

> you have 6.1.1 and 6.1.3 side-by-side, they should share the update information and inform the user to update to 6.1.4, for example.

For a 6.1.1 `pwsh` instance, the 6.1.2 and 6.1.3 are updates that should be shown to the user as a notification, to a 6.1.3 `pwsh` instance, that's not an update. So they don't share the update info 100%. It's cleaner to just let them use different folders. Also, practically it would be rare for someone to have many versions of `pwsh` installed, so it won't result in many folders.